certora/specs/AssetsAccounting.spec

```
@@ -0,0 +1,126 @@
+    // Assume no supply position to begin with.
+    require supplyShares(id, onBehalf) == 0;
+
+    uint256 suppliedAssets;
+    suppliedAssets, _ = supply(e, marketParams, assets, shares, onBehalf, data);
+
+    uint256 ownedAssets = ownedSupplyAssets(id, onBehalf);
+
+    assert suppliedAssets >= ownedAssets;
+}
+
+// Check that the assets withdrawn are less than the assets owned initially.
+rule withdrawAssetsAccounting(env e, MorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, address receiver) {
+    MorphoHarness.Id id = libId(marketParams);
+
+    // Assume no interest as it would increase the total supply assets.
+    require lastUpdate(id) == e.block.timestamp;
+
+    uint256 ownedAssets = ownedSupplyAssets(id, onBehalf);
+
+    uint256 withdrawnAssets;
+    withdrawnAssets, _ = withdraw(e, marketParams, assets, shares, onBehalf, receiver);
+
+    assert withdrawnAssets <= ownedAssets;
+}
+
+// Check that the assets borrowed are less than the assets owed in the end.
+rule borrowAssetsAccounting(env e, MorphoHarness.MarketParams marketParams, uint256 shares, address onBehalf, address receiver) {
+    MorphoHarness.Id id = libId(marketParams);
+
+    // Assume no interest as it would increase the total borrowed assets.
+    require lastUpdate(id) == e.block.timestamp;
+    // Assume no outstanding debt to begin with.
+    require borrowShares(id, onBehalf) == 0;
+
+    // The borrow call is restricted to shares as input to make it easier on the prover.
+    uint256 borrowedAssets;
+    borrowedAssets, _ = borrow(e, marketParams, 0, shares, onBehalf, receiver);
+
+    uint256 owedAssets = owedBorrowAssets(id, onBehalf);
+
+    assert borrowedAssets <= owedAssets;
+}
+
+// Check that the assets repaid are greater than the assets owed initially.
+rule repayAssetsAccounting(env e, MorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, bytes data) {
+    MorphoHarness.Id id = libId(marketParams);
+
+    // Assume no interest as it would increase the total borrowed assets.
+    require lastUpdate(id) == e.block.timestamp;
+
+    uint256 owedAssets = owedBorrowAssets(id, onBehalf);
+
+    uint256 repaidAssets;
+    repaidAssets, _ = repay(e, marketParams, assets, shares, onBehalf, data);
+
+    // Assume a full repay.
+    require borrowShares(id, onBehalf) == 0;
+
+    assert repaidAssets >= owedAssets;
+}
+
+// Check that the collateral assets supplied are greater than the assets owned in the end.
+rule supplyCollateralAssetsAccounting(env e, MorphoHarness.MarketParams marketParams, uint256 suppliedAssets, address onBehalf, bytes data) {
+    MorphoHarness.Id id = libId(marketParams);
+
+    // Assume no collateral to begin with.
+    require collateral(id, onBehalf) == 0;
+
+    supplyCollateral(e, marketParams, suppliedAssets, onBehalf, data);
+
+    uint256 ownedAssets = collateral(id, onBehalf);
+
+    assert suppliedAssets >= ownedAssets;
```

do we want to keep the general rule?
```suggestion
    assert suppliedAssets == ownedAssets;
```

---

Ok to have something even more precise :+1: